File: backend/src/main/java/com/evomind/entity/SocraticDialogue.java
package com.evomind.entity;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.time.LocalDateTime;

/**
 * 苏格拉底式对话会话实体
 * 记录用户与AI的苏格拉底式对话
 */
@Entity
@Table(name = "socratic_dialogues")
@Getter
@Setter
public class SocraticDialogue extends BaseEntity {

    @Column(name = "discussion_id", nullable = false)
    private Long discussionId; // 关联的讨论主题ID

    @Column(name = "user_id", nullable = false)
    private Long userId;

    @Column(nullable = false, length = 20)
    @Enumerated(EnumType.STRING)
    private DialogueStatus status = DialogueStatus.IN_PROGRESS;

    @Column(name = "current_round")
    private Integer currentRound = 0; // 当前对话轮次

    @Column(name = "max_rounds")
    private Integer maxRounds = 5; // 最大对话轮次

    @Column(name = "initial_question", columnDefinition = "TEXT")
    private String initialQuestion; // 初始问题

    @Column(name = "final_insight", columnDefinition = "TEXT")
    private String finalInsight; // 最终洞察总结

    @Column(name = "insight_generated_at")
    private LocalDateTime insightGeneratedAt;

    @Column(name = "user_satisfaction")
    private Integer userSatisfaction; // 用户满意度评分 1-5

    @Column(name = "is_abandoned")
    private Boolean isAbandoned = false; // 是否被用户放弃

    @Column(name = "abandoned_at")
    private LocalDateTime abandonedAt;

    @Column(name = "completed_at")
    private LocalDateTime completedAt;

    @Column(name = "last_message_at")
    private LocalDateTime lastMessageAt;

    @Column(name = "total_messages")
    private Integer totalMessages = 0;

    @Version
    private Long version; // 乐观锁，防止并发修改

    /**
     * 对话状态枚举
     */
    public enum DialogueStatus {
        IN_PROGRESS("进行中"),
        COMPLETED("已完成"),
        ABANDONED("已放弃"),
        INSIGHT_GENERATED("洞察已生成");

        private final String description;

        DialogueStatus(String description) {
            this.description = description;
        }

        public String getDescription() {
            return description;
        }
    }

    /**
     * 进入下一轮对话
     */
    public void nextRound() {
        this.currentRound++;
        this.lastMessageAt = LocalDateTime.now();
    }

    /**
     * 检查是否达到最大轮次
     */
    public boolean hasReachedMaxRounds() {
        return this.currentRound >= this.maxRounds;
    }

    /**
     * 标记对话完成
     */
    public void markCompleted() {
        this.status = DialogueStatus.COMPLETED;
        this.completedAt = LocalDateTime.now();
    }

    /**
     * 标记对话放弃
     */
    public void markAbandoned() {
        this.status = DialogueStatus.ABANDONED;
        this.isAbandoned = true;
        this.abandonedAt = LocalDateTime.now();
    }

    /**
     * 标记洞察已生成
     */
    public void markInsightGenerated(String insight) {
        this.finalInsight = insight;
        this.insightGeneratedAt = LocalDateTime.now();
        this.status = DialogueStatus.INSIGHT_GENERATED;
    }

    /**
     * 增加消息计数
     */
    public void incrementMessageCount() {
        this.totalMessages++;
    }

    /**
     * 检查是否可以继续对话
     */
    public boolean canContinue() {
        return this.status == DialogueStatus.IN_PROGRESS && !hasReachedMaxRounds();
    }
}

File: android-app/app/src/main/java/com/evomind/domain/model/SocraticDialogue.kt
package com.evomind.domain.model

import java.time.LocalDateTime

/**
 * 苏格拉底式对话会话
 */
data class SocraticDialogue(
    val id: Long,
    val discussionId: Long,
    val discussionTitle: String?,
    val status: DialogueStatus,
    val currentRound: Int,
    val maxRounds: Int,
    val initialQuestion: String?,
    val finalInsight: String?,
    val totalMessages: Int,
    val canContinue: Boolean,
    val remainingRounds: Int,
    val lastMessageAt: LocalDateTime?,
    val createdAt: LocalDateTime,
    val lastMessage: SocraticMessage?
) {
    companion object {
        const val DEFAULT_MAX_ROUNDS = 5
        const val MAX_MESSAGE_LENGTH = 2000
    }

    /**
     * 获取进度百分比
     */
    fun getProgressPercent(): Int {
        return if (maxRounds > 0) {
            (currentRound.toFloat() / maxRounds * 100).toInt()
        } else {
            0
        }
    }

    /**
     * 是否已完成
     */
    fun isCompleted(): Boolean {
        return status == DialogueStatus.COMPLETED || status == DialogueStatus.INSIGHT_GENERATED
    }

    /**
     * 是否已放弃
     */
    fun isAbandoned(): Boolean {
        return status == DialogueStatus.ABANDONED
    }

    /**
     * 是否进行中
     */
    fun isInProgress(): Boolean {
        return status == DialogueStatus.IN_PROGRESS
    }

    /**
     * 获取状态描述
     */
    fun getStatusDescription(): String {
        return when (status) {
            DialogueStatus.IN_PROGRESS -> "进行中 (${currentRound}/${maxRounds}轮)"
            DialogueStatus.COMPLETED -> "已完成"
            DialogueStatus.ABANDONED -> "已放弃"
            DialogueStatus.INSIGHT_GENERATED -> "洞察已生成"
        }
    }

    /**
     * 获取简短状态描述
     */
    fun getShortStatus(): String {
        return when (status) {
            DialogueStatus.IN_PROGRESS -> "进行中"
            DialogueStatus.COMPLETED -> "已完成"
            DialogueStatus.ABANDONED -> "已放弃"
            DialogueStatus.INSIGHT_GENERATED -> "已完成"
        }
    }
}

/**
 * 对话状态枚举
 */
enum class DialogueStatus(val description: String) {
    IN_PROGRESS("进行中"),
    COMPLETED("已完成"),
    ABANDONED("已放弃"),
    INSIGHT_GENERATED("洞察已生成");

    companion object {
        fun fromString(value: String): DialogueStatus {
            return values().find { it.name == value } ?: IN_PROGRESS
        }
    }
}

/**
 * 对话统计
 */
data class DialogueStats(
    val totalDialogues: Long,
    val completedDialogues: Long,
    val inProgressDialogues: Long,
    val averageRounds: Double,
    val totalMessages: Long,
    val averageDepthLevel: Double?
) {
    /**
     * 完成率
     */
    fun getCompletionRate(): Float {
        return if (totalDialogues > 0) {
            completedDialogues.toFloat() / totalDialogues
        } else {
            0f
        }
    }

    /**
     * 深度等级显示
     */
    fun getDepthLevelDisplay(): String {
        return averageDepthLevel?.let {
            String.format("%.1f", it)
        } ?: "-"
    }
}
